File: d5/java/d5b.java
import java.io.IOException;
import java.io.File;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

public class d5b {
    // not safe, buffer size converted from long to int
    // next time BufferedReader;)
    private static ArrayList<Byte> readPolymer(String fileName) {
        final File file = new File(fileName);
        final long size = file.length();
        ArrayList<Byte> array = new ArrayList<>((int) size);
        char[] buffer = new char[(int) size];
        try {
            final FileReader freader = new FileReader(file);
            freader.read(buffer);
            freader.close();
        } catch (IOException e) {
            System.out.println(e.getMessage());
        }
        for (int i = 0; i < size; i++) {
            array.add((byte) buffer[i]);
        }
        return array;
    }

    private static boolean react(byte a, byte b) {
        int diff = a - b;
        return (diff == 32 || diff == -32);
    }
    //TODO
    private static long reaction(final ArrayList<Byte> inputPolymer) {
        ArrayList<Byte> polyA = new ArrayList<Byte>(inputPolymer.size());
        ArrayList<Byte> polyB = new ArrayList<Byte>(inputPolymer);
        ArrayList<Byte> temp = null;

        do {
            temp = polyA;
            polyA = polyB;
            polyB = temp;
            polyB.clear();

            int i;
            for (i = 0; i < polyA.size() - 1; i++) {
                if (react(polyA.get(i), polyA.get(i + 1))) ++i;
                else {
                    polyB.add(polyA.get(i));
                }
            } 
            if (i == polyA.size() - 1) polyB.add(polyA.get(i));
        } while (polyA.size() != polyB.size());
        return polyA.size();
    }

    private static ArrayList<Byte> filter(ArrayList<Byte> lst, byte elem) {
        final ArrayList<Byte> in = lst;
        ArrayList<Byte> out = new ArrayList<>();
        for (byte c : in) {
            if (c != elem && c != elem + 32) out.add(c);
        }
        return out;
    }

    public static void main(String[] args) {
        long answer = 0;

        long start = System.nanoTime();
        final ArrayList<Byte> polymer = readPolymer("input.txt");
        final ArrayList<Character> elements = new ArrayList<>(Arrays.asList('A', 'B', 'C', 'D', 'E',
             'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'));
        ArrayList<Long> lens = new ArrayList<>();

        for (char elem : elements) {
            lens.add(reaction(filter(polymer, (byte) elem)));
        }
        answer = Collections.min(lens);
        long end = System.nanoTime() - start;

        System.out.println("Answer:   " + Long.toString(answer));
        System.out.println("time:   " + end / 1000000 + " ms");
    }
}

File: d5/py/d5b.py
from time import perf_counter_ns
import profile

def readPolymer(fileName):
    file = open(fileName)
    input = file.read()
    polymer = [ord(x) for x in input]
    return polymer

def reaction(polymer):
    polyB = polymer
    polyA = []

    reactions = 1
    while reactions > 0:
        polyA = polyB
        polyB = []
        reactions = 0

        i = 0
        sz = len(polyA)
        while i < sz - 1:
            if _react(polyA[i], polyA[i + 1]):
                i += 2
                reactions += 1
            else:
                polyB.append(polyA[i])
                i += 1
        if i == sz - 1:
            polyB.append(polyA[i])
                
    return len(polyA)

def _react(a, b):
    return a - b == 32 or a - b == -32

def main():
    answer = 0

    timing = perf_counter_ns()
    polymer = readPolymer("input.txt")
    lens = []
    elements = ['A', 'B', 'C', 'D', 'E',
             'F', 'G', 'H', 'I', 'J', 'K',
             'L', 'M', 'N', 'O', 'P', 'Q',
             'R', 'S', 'T', 'U', 'V', 'W',
              'X', 'Y', 'Z']
   
    for c in [ord(x) for x in elements]:
        lens.append(reaction([x for x in polymer if x != c and x != c + 32]))

    answer = min(lens)
    timing = perf_counter_ns() - timing

    print("Answer:  " + str(answer))
    print("time:  " + str(timing // 1000000) + " ms")
    return

if __name__ == "__main__": main()